File: trunk/bcore/file_info.cpp
/*
 * btg Copyright (C) 2005 Michael Wojciechowski.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/*
 * $Id$
 */

#include <bcore/file_info.h>
#include <bcore/t_string.h>
#include <bcore/helpermacro.h>

namespace btg
{
   namespace core
   {
      fileInformation::fileInformation()
         : Serializable(),
           Printable(),
           filename("undefined file"),
           pieces_size(0),
           pieces(0),
           bytesPerPiece(0),
           fileSize(0),
           status(fileInformation::DONE_UNDEF)
      {
      }


      fileInformation::fileInformation(fileInformation const& _fileinfo)
         : Serializable(),
           Printable(),
           filename(_fileinfo.filename),
           pieces_size(_fileinfo.pieces_size),
           pieces(_fileinfo.pieces),
           bytesPerPiece(_fileinfo.bytesPerPiece),
           fileSize(_fileinfo.fileSize),
           status(_fileinfo.status)
      {
      }

      fileInformation::fileInformation(std::string const& _filename,
                                       t_int const _bytesPerPiece,
                                       t_ulong const _file_size)
         : Serializable(),
           Printable(),
           filename(_filename),
           pieces_size(0),
           pieces(0),
           bytesPerPiece(_bytesPerPiece),
           fileSize(_file_size),
           status(fileInformation::DONE_ALL)
      {
         
      }

      fileInformation::fileInformation(std::string const& _filename, 
                                       t_bitVector const& _pieces, 
                                       t_int const _bytesPerPiece, 
                                       t_ulong const _file_size)
         : Serializable(),
           Printable(),
           filename(_filename),
           pieces_size(0),
           pieces(0),
           bytesPerPiece(_bytesPerPiece),
           fileSize(_file_size),
           status(fileInformation::DONE_UNDEF)
      {
         if (_pieces.size() > 0)
            {
               t_bitVectorCI iter;
               t_uint b_set = 0;

               for (iter = _pieces.begin(); 
                    iter != _pieces.end();
                    iter++)
                  {
                     if (*iter)
                        {
                           b_set++;
                        }
                  }

               if (b_set == _pieces.size())
                  {
                     status      = fileInformation::DONE_ALL;
                     pieces_size = _pieces.size();
                  }
               else if (b_set == 0)
                  {
                     status      = fileInformation::DONE_NONE;
                     pieces_size = _pieces.size();
                  }
               else
                  {
                     status      = fileInformation::DONE_UNDEF;
                     pieces      = _pieces;
                     pieces_size = _pieces.size();
                  }
            } // pieces size is not 0.
         else
            {
               status      = fileInformation::DONE_UNDEF;
               pieces_size = 0;
            }
      }

      bool fileInformation::operator== (fileInformation const& _fileinfo) const
      {
         if (this->filename != _fileinfo.filename)
            {
               return false;
            }

         if (this->pieces_size != _fileinfo.pieces_size)
            {
               return false;
            }

         if (this->status != _fileinfo.status)
            {
               return false;
            }

         if (this->pieces != _fileinfo.pieces)
            {
               return false;
            }

         if (this->bytesPerPiece != _fileinfo.bytesPerPiece)
            {
               return false;
            }

         if (this->fileSize != _fileinfo.fileSize)
            {
               return false;
            }

         return true;
      }

      bool fileInformation::serialize(btg::core::externalization::Externalization* _e) const
      {
         // string, filename;
         _e->stringToBytes(&this->filename);
         BTG_RCHECK(_e->status());

         t_int temp_status = this->status;
         _e->intToBytes(&temp_status);
         BTG_RCHECK(_e->status());

         _e->intToBytes(&this->bytesPerPiece);
         BTG_RCHECK(_e->status());

         _e->uLongToBytes(&this->fileSize);
         BTG_RCHECK(_e->status());

         // Length, in bits.
         _e->intToBytes(&this->pieces_size);
         BTG_RCHECK(_e->status());

         if (status == fileInformation::DONE_UNDEF)
         {
            pieces.serialize(_e);
            BTG_RCHECK(_e->status());
         }

         return true;
      }

      bool fileInformation::deserialize(btg::core::externalization::Externalization* _e)
      {
         // string, filename;
         _e->bytesToString(&this->filename);
         BTG_RCHECK(_e->status());

         t_int temp_status = 0;
         _e->bytesToInt(&temp_status);
         BTG_RCHECK(_e->status());
         this->status = (bitsetModifier)temp_status;

         _e->bytesToInt(&this->bytesPerPiece);
         BTG_RCHECK(_e->status());

         _e->bytesToULong(&this->fileSize);
         BTG_RCHECK(_e->status());

         // Length, in bits.
         _e->bytesToInt(&this->pieces_size);
         BTG_RCHECK(_e->status());

         if (status == fileInformation::DONE_UNDEF)
         {
            pieces.deserialize(_e);
            BTG_RCHECK(_e->status());
         }

         return true;
      }

      t_uint fileInformation::size() const
      {
         return pieces_size;
      }

      std::string fileInformation::getFilename() const
      {
         return filename;
      }

      t_bitVector fileInformation::getBits() const
      {
         if ((this->status != fileInformation::DONE_ALL) && 
             (this->status != fileInformation::DONE_NONE))
            {
               return pieces;
            }

         t_bitVector list_to_return;

         if (this->status == fileInformation::DONE_ALL)
            {
               for (t_int i=0; i<this->pieces_size; i++)
                  {
                     list_to_return.push_back(1);
                  }
            }
         else if (this->status == fileInformation::DONE_NONE)
            {
               for (t_int i=0; i<this->pieces_size; i++)
                  {
                     list_to_return.push_back(0);
                  }
            }

         return list_to_return;
      }

      std::string fileInformation::toString() const
      {
         t_int const lineWidth = 79;

         std::string output;

         output += "Filename: ";
         output += this->filename;
         output += GPD->sNEWLINE();

         output += "Size: ";
         output += convertToString<t_uint>(this->pieces_size);
         output += GPD->sNEWLINE();

         output += "Bytes per piece: ";
         output += convertToString<t_uint>(this->bytesPerPiece);
         output += GPD->sNEWLINE();



         switch (this->status)
            {
            case fileInformation::DONE_ALL:
               {
                  output += "All bits set.";
                  output += GPD->sNEWLINE();
                  break;
               }
            case fileInformation::DONE_NONE:
               {
                  output += "All bits unset.";
                  output += GPD->sNEWLINE();
                  break;
               }
            case fileInformation::DONE_UNDEF:
               {
                  if (this->pieces_size <= 0)
                     {
                        output += "Undefined.";
                        output += GPD->sNEWLINE();
                     }
                  else
                     {
                        t_uint const max_char = lineWidth;
                        t_uint charcount      = 0;
                        t_bitVectorCI iter;

                        for (iter = pieces.begin(); iter != pieces.end(); iter++)
                           {
                              if (*iter)
                                 {
                                    output += "X";
                                 }
                              else
                                 {
                                    output += "_";
                                 }

                              charcount++;
                              if (charcount >= max_char)
                                 {
                                    charcount  = 0;
                                    output    += GPD->sNEWLINE();
                                 }
                           } // for iter

                     } // else
                  break;
               }
            } // switch

         return output;
      }

      bool fileInformation::isFull() const
      {
         bool op_status = false;
         if (this->status == fileInformation::DONE_ALL)
            {
               op_status = true;
            }
         return op_status;
      }

      bool fileInformation::isEmpty() const
      {
         bool op_status = false;
         if (this->status == fileInformation::DONE_NONE)
            {
               op_status = true;
            }
         return op_status;
      }

      t_ulong fileInformation::getFileSize() const
      {
         return fileSize;
      }

      fileInformation::~fileInformation()
      {
      }

      /* */

      selectedFileEntry::selectedFileEntry()
         : Serializable(),
           filename_("none"),
           selected_(true)
      {
      }

      selectedFileEntry::selectedFileEntry(std::string const& _filename)
         : Serializable(),
           filename_(_filename),
           selected_(true)
      {
      }

      selectedFileEntry::selectedFileEntry(std::string const& _filename,
                                           bool const _selected)
         : Serializable(),
           filename_(_filename),
           selected_(_selected)
      {

      }

      bool selectedFileEntry::serialize(btg::core::externalization::Externalization* _e) const
      {
         BTG_RCHECK( _e->stringToBytes(&this->filename_) );
         BTG_RCHECK( _e->boolToBytes(this->selected_) );
         return true;
      }

      bool selectedFileEntry::deserialize(btg::core::externalization::Externalization* _e)
      {
         BTG_RCHECK( _e->bytesToString(&this->filename_) );
         BTG_RCHECK( _e->bytesToBool(this->selected_) );

         return true;
      }

      void selectedFileEntry::select()
      {
         selected_ = true;
      }

      void selectedFileEntry::deSelect()
      {
         selected_ = false;
      }

      std::string selectedFileEntry::filename() const
      {
         return filename_;
      }

      bool selectedFileEntry::selected() const
      {
         return selected_;
      }

      selectedFileEntry::selectedFileEntry(selectedFileEntry const& _entry)
         : Serializable(),
           filename_(_entry.filename_),
           selected_(_entry.selected_)
      {
      }

      bool selectedFileEntry::operator== (selectedFileEntry const& _entry) const
      {
         if (
             (_entry.filename_ == filename_) &&
             (_entry.selected_ == selected_)
             )
            {
               return true;
            }

         return false;
      }

      selectedFileEntry& selectedFileEntry::operator= (selectedFileEntry const& _entry)
      {
         bool eq = (_entry == *this);
         if (!eq)
            {
               filename_ = _entry.filename_;
               selected_ = _entry.selected_;
            }

         return *this;
      }

      selectedFileEntry::~selectedFileEntry()
      {
         
      }

      /* */

      selectedFileEntryList::selectedFileEntryList()
         : Serializable(),
           files_()
      {
      }

      selectedFileEntryList::selectedFileEntryList(std::vector<selectedFileEntry> const& _files)
         : Serializable(),
           files_(_files)
      {

      }

      selectedFileEntryList::selectedFileEntryList(selectedFileEntryList const& _list)
         : Serializable(),
           files_(_list.files_)
      {
      }

      void selectedFileEntryList::add(std::string const& _filename, 
                                      bool _selected)
      {
         selectedFileEntry sfe(_filename, _selected);
         files_.push_back(sfe);
      }

      bool selectedFileEntryList::operator== (selectedFileEntryList const& _list) const
      {
         if (_list.files_ == files_)
            {
               return true;
            }

         return false;
      }

      selectedFileEntryList& selectedFileEntryList::operator= (selectedFileEntryList const& _list)
      {
         bool eq = (_list == *this);
         if (!eq)
            {
               files_ = _list.files_;
            }

         return *this;
      }

      bool selectedFileEntryList::serialize(btg::core::externalization::Externalization* _e) const
      {
         t_int const list_size = files_.size();

         // Write the length of the list:
         BTG_RCHECK( _e->intToBytes(&list_size) );

         // Write each entry.
         std::vector<selectedFileEntry>::const_iterator iter;
         for (iter = files_.begin();
              iter != files_.end();
              iter++)
            {
               BTG_RCHECK( iter->serialize(_e) );
            }
         
         return true;
      }

      bool selectedFileEntryList::deserialize(btg::core::externalization::Externalization* _e)
      {
         t_int list_size = 0;

         // First, get the size of the list:
         BTG_RCHECK( _e->bytesToInt(&list_size) );

         selectedFileEntry sfe;
         for (t_int i=0; i<list_size; i++)
            {
               BTG_RCHECK( sfe.deserialize(_e) );
               files_.push_back(sfe);
            }

         return true;
      }

      void selectedFileEntryList::select(std::string const& _filename)
      {
         setFlag(_filename, true);
      }

      void selectedFileEntryList::deSelect(std::string const& _filename)
      {
         setFlag(_filename, false);
      }
      
      std::vector<selectedFileEntry> selectedFileEntryList::files() const
      {
         return files_;
      }

      void selectedFileEntryList::setFiles(std::vector<selectedFileEntry> const& _files)
      {
         files_ = _files;
      }

      void selectedFileEntryList::setFlag(std::string const& _filename, 
                                          bool const _flagValue)
      {
         std::vector<selectedFileEntry>::iterator iter;
         for (iter = files_.begin();
              iter != files_.end();
              iter++)
            {
               if (iter->filename() == _filename)
                  {
                     break;
                  }
            }

         if (iter != files_.end())
            {
               if (_flagValue)
                  {
                     iter->select();
                  }
               else
                  {
                     iter->deSelect();
                  }
            }
      }

      t_int selectedFileEntryList::size() const
      {
         return files_.size();
      }

      bool selectedFileEntryList::selected(std::string const& _filename) const
      {
         std::vector<selectedFileEntry>::const_iterator iter;
         for (iter = files_.begin();
              iter != files_.end();
              iter++)
            {
               if (iter->filename() == _filename)
                  {
                     break;
                  }
            }

         if (iter != files_.end())
            {
               if (iter->selected())
                  {
                     return true;
                  }
            }

         return false;
      }

      selectedFileEntryList::~selectedFileEntryList()
      {

      }

      /* */

   } // namespace core
} // namespace btg

File: tags/0.9.8/UI/www/php/htdocs/bcore/file_info.php
<?php
/*
 * btg Copyright (C) 2005 Michael Wojciechowski.
 * PHP client written by Johan Ström.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/*
 * $Id: file_info.php,v 1.1.2.1 2006/03/27 07:17:08 jstrom Exp $
 */
if(!defined("BTG_BCORE_PATH"))
	define("BTG_BCORE_PATH", "bcore");

require_once(BTG_BCORE_PATH."/serializable.php");

class selectedFileEntry extends BTGSerializable
{
   private $filename;
   private $selected;

   public function selectedFileEntry($filename = "", $selected = false)
   {
      $this->filename = $filename;
      $this->selected = $selected;
   }

   public function select()
   {
      $this->selected = true;
   }

   public function deSelect()
   {
      $this->selected = false;
   }

   public function getFilename()
   {
      return $this->filename;
   }

   public function getSelected()
   {
      return $this->selected;
   }

   public function serialize(&$a = array())
	{
      $this->stringToBytes($a, $this->filename);
		$this->boolToBytes($a, $this->selected);

      return $a;
   }

   public function deserialize(&$data)
   {
      $this->bytesToString($this->filename, $data);
		$this->bytesToBool($this->selected, $data);
   }
}

class fileInformation extends BTGSerializable
{
	const MAX_PASS_COUNT = 5;
	const DONE_ALL = 1;
	const DONE_NONE = 2;
	const DONE_UNDEF = 0;

	const BITS_PER_BYTE = 7;
	/// The filename this class holds piece-info about.
	private $filename;
	/// The pieces of this file.
	private $pieces;
	/// The number of pieces.
	private $pieces_size;
	/// The number of bytes each piece represents.
	private $bytesPerPiece;
	/// The exact file size.
	private $fileSize;
	/// Status
	private $status;

	/// Constructor, $pieces should be an array of booleans
	public function fileInformation($filename="undefined file", $pieces=array(), $bytesPerPiece=0, $file_size=0)
	{
		$this->filename = $filename;
		$this->bytesPerPiece = $bytesPerPiece;
		$this->fileSize = $file_size;
		$this->status = fileInformation::DONE_UNDEF;

		if(count($pieces) > 0)
		{
			$b_set = 0;
			foreach($pieces as $piece)
			{
				if($piece == true)
					$b_set++;
			}

			if($b_set == count($pieces))
				$this->status = fileInformation::DONE_ALL;
			else if($b_set == 0)
				$this->status = fileInformation::DONE_NONE;
			else
			{
				$this->status = fileInformation::DONE_UNDEF;
				$this->pieces = $pieces;
			}
		}else
		{
			$this->status = fileInformation::DONE_UNDEF;
		}

		$this->pieces_size = count($pieces);
	}

	/// Return the number of pieces this class holds.
	public function size()
	{
		return $this->pieces_size;
	}

	/// Return the exact file size.
	public function getFileSize()
	{
		return $this->fileSize;
	}

	/// Return the filename
	public function getFilename()
	{
		return $this->filename;
	}

	/// Return the bits.
	public function getBits()
	{
		if(($this->status != fileInformation::DONE_ALL) && ($this->status != fileInformation::DONE_NONE))
			return $this->pieces;

		$ret = array();
		if($this->status == fileInformation::DONE_ALL)
		{
			for($i = 0; $i < $this->pieces_size; $i++)
				$ret[] = true;
		}
		else if($this->status == fileInformation::DONE_NONE)
		{
			for($i = 0; $i < $this->pieces_size; $i++)
				$ret[] = false;
		}

		return $ret;
	}

	/// Returns true if all bits are set.
	public function isFull()
	{
		if($this->status == fileInformation::DONE_ALL)
			return true;
		return false;
	}

	/// Returns true if none of the contained bits are set.
	public function isEmpty()
	{
		if($this->status == fileInformation::DONE_NONE)
			return true;
		return false;
	}

	public function serialize(&$a = array())
	{
		$this->stringToBytes($a, $this->filename);
		$this->intToBytes($a, $this->status);
		$this->intToBytes($a, $this->bytesPerPiece);
		$this->uLongToBytes($a, $this->fileSize);
		$this->intToBytes($a, $this->pieces_size);

		if($this->status == fileInformation::DONE_UNDEF)
		{
			$bit_counter = 0;
			$output = 0;
			$data = "";

			foreach($this->pieces as $piece)
			{
				if($bit_counter >= fileInformation::BITS_PER_BYTE)
				{
					$data .= chr($output);
					$output = 0;
					$bit_counter = 0;
				}

				if($piece)
					$output|= 1 << $bit_counter;

				$bit_counter++;
			}

			if($bit_counter > 0)
			{
				$data .= chr($output);
			}
			
			$this->intToBytes($a, $bit_counter);
			$this->addBytes($a, $data, strlen($data));
		}
		return $a;
	}

	public function deserialize(&$data)
	{
		$this->bytesToString($this->filename, $data);
		$this->bytesToInt($this->status, $data);
		$this->bytesToInt($this->bytesPerPiece, $data);
		$this->bytesToULong($this->fileSize, $data);
		$this->bytesToInt($this->pieces_size, $data);
		
		if($this->status == fileInformation::DONE_UNDEF)
		{
			$bytes = "";
			$bits = 0;
			$octets = 0;
			$this->bytesToInt($bits, $data);
			$this->getBytes($bytes, $data, $octets);

			$this->pieces = array();

			$remainder =  $octets * fileInformation::BITS_PER_BYTE - $bits;
			$octets -= $remainder ? 1 : 0;

			if($octets > 0)
			{
				for($i=0; $i < $octets; $i++)
				{
					$b = $bytes[$i];
					for($n=0; $n < fileInformation::BITS_PER_BYTE; $n++)
					{
						if( (($b >> $n) & 1) == 1 )
							$this->pieces[] = true;
						else
							$this->pieces[] = false;
					}
				}
			}

			if($remainder > 0)
			{
				$b = $bytes[$octets];
				for($n = 0; $n < $remainder; $n++)
				{
					if( (($b >> $n) & 1) == 1 )
						$this->pieces[] = true;
					else
						$this->pieces[] = false;
				}
			}
		}
	}

	public function toString($lineWidth = 80)
	{
		$ret = "Filename: ". $this->filename . "\n";
		$ret.= "Pieces: ". $this->pieces_size ."\n";
		$ret.= "Bytes per piece: ". $this->bytesPerPiece ."\n";

		switch($this->status)
		{
		case fileInformation::DONE_ALL:
			$ret.= "All bits set.\n";
			break;
		case fileInformation::DONE_NONE:
			$ret.= "All bits unset.\n";
			break;
		case fileInformation::DONE_UNDEF:
			if($this->pieces_size <= 0)
				$ret.= "Undefined.\n";
			else
			{
				$charcount = 0;
				foreach($this->pieces as $piece)
				{
					if($piece)
						$ret.= "X";
					else
						$ret.= "_";
					
					$charcount++;
					if($charcount >= $lineWidth)
					{
						$charcount = 0;
						$ret.= "\n";
					}
				}
				$ret.="\n";
			}
			break;
		}

		return $ret;
	}
}
?>
